### PR TITLE
Add worker-loader

### LINKED ISSUE #814: scriptURL of serviceWorker.register isn't considered part of bundle
👋🏼 

I'm currently working with microbundle to package a service worker library. I have a file [`index.mjs`](https://github.com/TimDaub/web3-service-worker-cache/blob/demo/src/index.mjs) that takes care of initializing and handling the SW in the main thread. And I have a file [`worker.mjs`](https://github.com/TimDaub/web3-service-worker-cache/blob/demo/src/worker.mjs) that separates the service worker code.

After browser compatbility, I register the worker like this:

```js
await navigator.serviceWorker.register('./worker.mjs');
```

I use `microbundle` in a standard way by configuring it in [package.json](https://github.com/TimDaub/web3-service-worker-cache/blob/demo/package.json).

I'm aware that no bundler would take the input to `register()` and make it somehow part of the output. Usually, a developer has to care for this separately e.g. in rollup by adding a secondary input and output. But today while doing that I found it annoying and I asked myself why?

Couldn't microbundle just take care of transpiling the worker.mjs file and include it somehow. E.g. it could be bundled as a base64 data URI...

### PR DESCRIPTION
**Pull request by TimDaub**
- Fixes #814 
- Superseeds #757

Few comments
- I tried rebasing on the latest master but I got confused about the package-lock.json file. I'd appreciate some help. I'm not sure how to solve all of this properly.
- I ended up using https://github.com/surma/rollup-plugin-off-main-thread which @developit recommended in #757 
- I didn't put TypeScript support because I don't know TypeScript. I've made a note in the readme that it only works with JS builds for now. Maybe someone can add TS in a separate PR...

🤞🏼 that somehow this gets merged!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -243,6 +261,7 @@ Options
 	--jsxImportSource  Specify the automatic import source for JSX like preact
 	--tsconfig         Specify the path to a custom tsconfig.json
 	--css-modules      Configures .css to be treated as modules (default: null)
+	--worker-loader    Generate inline worker blobs  (default false)
```
Comment (developit) [inline README.md:264]:
```suggestion
	--workers          Bundle module workers - see https://git.io/J3oSF  (default false)
```

File: src/index.js [hunk 2]
```diff
@@ -366,6 +367,7 @@ function createConfig(options, entry, format, writeMeta) {
 			: () => resolve(options.cwd, 'mangle.json');
 
 	const useTypescript = extname(entry) === '.ts' || extname(entry) === '.tsx';
+	const useWorkerLoader = options['worker-loader'] !== false;
```
Comment (developit) [inline src/index.js:370]:
```suggestion
	const useWorkerLoader = options.workers !== false;
```

File: src/prog.js [hunk 3]
```diff
@@ -56,6 +56,7 @@ export default handler => {
 		.option('--name', 'Specify name exposed in UMD builds')
 		.option('--cwd', 'Use an alternative working directory', '.')
 		.option('--sourcemap', 'Generate source map', true)
+		.option('--worker-loader', 'Generate inline worker blobs', false)
```
Comment (developit) [inline src/prog.js:59]:
```suggestion
		.option('--workers', 'Bundle module workers - see https://git.io/J3oSF', false)
```

File: test/__snapshots__/index.test.js.snap [hunk 4]
```diff
@@ -2745,3 +2745,279 @@ exports[`fixtures build ts-module with microbundle 7`] = `
 //# sourceMappingURL=ts-module.umd.js.map
 "
 `;
+
+exports[`fixtures build worker-loader with microbundle 1`] = `
+"Used script: microbundle -f modern,es --worker-loader
```
Comment (developit) [inline test/__snapshots__/index.test.js.snap:2750]:
```suggestion
"Used script: microbundle -f modern,es --workers
```

File: test/__snapshots__/index.test.js.snap [hunk 5]
```diff
@@ -2745,3 +2745,279 @@ exports[`fixtures build ts-module with microbundle 7`] = `
...
+      CHANGELOG.md
+      README.md
+      dist
+        tippex.es.js
+        tippex.es.js.map
+        tippex.umd.js
+        tippex.umd.js.map
+      package.json
+      src
+        index.js
```
Comment (developit) [inline test/__snapshots__/index.test.js.snap:2979]: not sure why node_modules is being included here - maybe forgot to delete it before running the fixture?

File: test/fixtures/worker-loader/package.json [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "worker-loader",
+  "scripts": {
+    "build": "microbundle -f modern,es --worker-loader"
```
Comment (developit) [inline test/fixtures/worker-loader/package.json:4]:
```suggestion
    "build": "microbundle -f modern,es --workers"
```

File: README.md [hunk 7]
```diff
@@ -192,6 +192,24 @@ By default Microbundle outputs multiple bundles, one bundle per format. A single
...
+
+```js
+const worker = new Worker('./worker.js', { type: 'module' });
+```
+
+And build it like this:
+
+```bash
+microbundle --worker-loader
+```
```
Comment (developit) [inline README.md:208]:
````suggestion
### Building Module Workers

Microbundle is able to detect and bundle Module Workers when generating bundles in the
`es`, `umd` and `modern` formats. To use this feature, instantiate your Web Worker as follows:

```js
worker = new Worker(new URL('./worker.js', import.meta.url), { type: 'module' });
// or simply:
worker = new Worker('./worker.js', { type: 'module' });
```

... then add the `--workers` flag to your build command:

```bash
microbundle --workers
```
````

File: test/__snapshots__/index.test.js.snap (hunk 5 above)
Comment (TimDaub) [reply test/__snapshots__/index.test.js.snap:2979]: oh, that could be. I'll take a look

File: src/prog.js (hunk 3 above)
Comment (TimDaub) [reply src/prog.js]: for some reason eslint doesn't like it. I don't understand its error message:

```
59:11  error  Replace `'--workers',·'Bundle·module·workers·-·see·https://git.io/J3oSF',·false` with `⏎↹↹↹'--workers',⏎↹↹↹'Bundle·module·workers·-·see·h
ttps://git.io/J3oSF',⏎↹↹↹false,⏎↹↹`  prettier/prettier
```

Comment (developit) [reply src/prog.js]: heh - it just wants the message to be broken across multiple lines (kinda silly in this case, but alas).

Comment (TimDaub) [reply src/prog.js]: OK, makes sense.

File: test/__snapshots__/index.test.js.snap (hunk 5 above)
Comment (TimDaub) [reply test/__snapshots__/index.test.js.snap]: Should be resolved in f155eee

**Comment by TimDaub** [comment]
@developit I addressed your comments. I'm not sure I'm capable of rebasing though. Last time I tried sketchy things happened in package-lock.json that I wasn't able to address. What now?

**Comment by ForsakenHarmony** [comment]
> Last time I tried sketchy things happened in package-lock.json that I wasn't able to address.

You generally have to accept the remote one and then after create a new commit / amend a commit with a file you generated after by running `npm install`

**Comment by developit** [comment]
I'll try to get this rebased this week (I'm also terrible at it!).

This will need to go into a new breaking release (minor since microbundle is still pre-1.0), but I think that should be fine.

**Comment by developit** [comment]
Just wanted to mention I haven't forgotten about this PR, I just have been side-tracked rewriting Microbundle's filename inference.

**Review by developit — APPROVED** [review]

**Comment by developit** [comment]
This was landed in #867! 🎉

**Pull request closed without merging**